backend/packages/Upgrade/src/api/services/SegmentService.ts

```diff
@@ -185,62 +182,71 @@ export class SegmentService {
     return await this.segmentRepository.deleteSegment(id, logger);
   }
 
-  public async importSegments(segments: SegmentFile[], logger: UpgradeLogger): Promise<SegmentReturnObj> {
+  public async validateSegments(segments: SegmentFile[], logger: UpgradeLogger): Promise<SegmentImportError[]> {
+    logger.info({ message: `Validating segments` });
+    const validatedSegments = await this.checkSegmentsValidity(segments);
+    const validationErrors = validatedSegments.importErrors.filter((seg) => seg.error !== null);
+    return validationErrors;
+  }
+
+  public async importSegments(segments: SegmentFile[], logger: UpgradeLogger): Promise<SegmentImportError[]> {
+    const validatedSegments = await this.checkSegmentsValidity(segments);
+    for (const segment of validatedSegments.segments) {
+      // Giving new id to avoid segment duplication
+      segment.id = uuid();
+
+      logger.info({ message: `Import segment => ${JSON.stringify(segment, undefined, 2)}` });
+      await this.addSegmentDataInDB(segment, logger);
+    }
+    return validatedSegments.importErrors;
+  }
+
+  public async checkSegmentsValidity(segments: SegmentFile[]): Promise<SegmentValidationObj> {
     const importFileErrors: SegmentImportError[] = [];
     const parsedData: SegmentInputValidator[] = [];
     const fileNames: string[] = [];
-    for (const segment of segments) {
-      const fileName = segment.fileName.split('.');
-      if (segment.fileName.endsWith('.json')) {
-        const segmentForValidation = this.convertJSONStringToSegInputValFormat(segment.fileContent);
-        const segmentJSONValidation = await this.validateObject(segmentForValidation);
-
-        if (segmentJSONValidation.isSegmentValid) {
-          fileNames.push(fileName[0]);
-          parsedData.push(segmentForValidation.segment);
-        } else {
-          importFileErrors.push({
-            fileName: fileName[0],
-            error: 'Invalid Segment JSON, missing properties: ' + segmentJSONValidation.missingProperty,
-          });
+
+    await Promise.all(
+      segments.map(async (segment) => {
+        const extension = path.extname(segment.fileName);
+        if (extension === '.json') {
+          let segmentForValidation = this.convertJSONStringToSegInputValFormat(segment.fileContent);
+          segmentForValidation = plainToClass(SegmentInputValidator, segmentForValidation);
+          const segmentJSONValidation = await this.checkForMissingProperties(segmentForValidation);
+          const fileName = segment.fileName.slice(0, segment.fileName.lastIndexOf('.'));
```

Comment:
This part of the code should be moved outside of the `promise.all` and you should use array's filter for creating an array instead of pushing data in the array.